src/java.net.http/share/classes/jdk/internal/net/http/common/HttpBodySubscriberWrapper.java

```diff
@@ -125,34 +172,139 @@ private void propagateError(Throwable t) {
         }
```

Comment:
I haven't fully wrapped my head around the possible flow of the user subscription, but is there a possiblity where this call to `onError(...)` here results in an reentrant call to this `propagateError()`? For that matter, not just reentrant but perhaps from a different thread concurrently? The reason I ask is, should we call this onError just once? The `java.util.concurrent.Flow.Subscriber.onError(Throwable)` method says this:
> Method invoked upon an unrecoverable error encountered by a
> Publisher or Subscription, after which no other Subscriber
> methods are invoked by the Subscription. 

So I'm wondering if we should maintain some state to only invoke it once?

Reply:
This is the purpose of this class: complete() and propagateError() should ensure that onError() is only called once in the wrapped subscriber. The markCompleted() call should ensure that, even if there is a reentrant call.

Reply:
You are right - the error propagation through `propagateError()` always happens through the `complete()` method of this wrapper class and that method has the necessary state management to call this only once.